lib/ex_aws/credentials_ini/file.ex

```diff
@@ -1,17 +1,111 @@
 if Code.ensure_loaded?(ConfigParser) do
   defmodule ExAws.CredentialsIni.File do
+    import ExAws.Request.Hackney, only: [request: 4]
     # as per https://docs.aws.amazon.com/cli/latest/topic/config-vars.html
     @valid_config_keys ~w(
       aws_access_key_id aws_secret_access_key aws_session_token region
       role_arn source_profile credential_source external_id mfa_serial role_session_name credential_process
+      sso_start_url sso_region sso_account_id sso_role_name
     )
 
     def security_credentials(profile_name) do
       shared_credentials = profile_from_shared_credentials(profile_name)
       config_credentials = profile_from_config(profile_name)
+      merge_credentials(config_credentials, shared_credentials)
+    end
+
+    defp merge_credentials(
+           %{
+             sso_start_url: sso_start_url,
+             sso_account_id: sso_account_id,
+             sso_role_name: sso_role_name,
+           } = config_credentials,
+           _shared_credentials
+         ) do
+           case get_sso_role_credentials(sso_start_url, sso_account_id, sso_role_name) do
+             {:ok, sso_creds} -> {:ok, Map.merge(sso_creds, config_credentials)}
+             {:error, _} = err -> err
+           end
+    end
+
+    defp merge_credentials(config_credentials, shared_credentials) do
       Map.merge(config_credentials, shared_credentials)
     end
 
+    defp get_sso_cache_file(sso_start_url) do
+      hash = :crypto.hash(:sha, sso_start_url) |> Base.encode16() |> String.downcase()
+
+      System.user_home()
+      |> Path.join(".aws/sso/cache/#{hash}.json")
+    end
+
+    defp get_sso_role_credentials(sso_start_url, sso_account_id, sso_role_name) do
+      sso_start_url
+      |> get_sso_cache_file()
+      |> File.read()
+      |> parse_sso_cache_file
+      |> check_sso_expiration
+      |> request_sso_role_credentials(sso_account_id, sso_role_name)
+      |> rename_sso_credential_keys
+    end
+
+    defp parse_sso_cache_file({:ok, contents}) do
+      contents
+      |> Jason.decode()
+    end
+
+    defp parse_sso_cache_file(err), do: err
+
+    defp check_sso_expiration({:ok, %{"expiresAt" => expires_at_str} = config}) do
+      with {:ok, expires_at, _} <- DateTime.from_iso8601(expires_at_str)
+      do
+        case DateTime.compare(expires_at, DateTime.utc_now()) do
+          :gt -> {:ok, config}
+          _ -> {:error, "SSO access token is expired, refresh the token with `aws sso login`"}
+        end
+      else
+        {:error, err} -> {:error, "SSO cache file has invalid expiration format: #{err}"}
+      end
+    end
+
+    defp check_sso_expiration(err), do: err
+
+    #TODO: Implement check that verifies if a user has aws_sso and source profile configured
+    # Saw this concern raised in saml2aws using aws go sdk and think it could be helpful
+
+    # TODO: the :ex_aws.request() abstraction that can choose hackney or what users set requires signing
```

Comment:
Did you have any thoughts here? This is where I hit a wall on the best approach last time I worked on this. IIRC `hackney` is not a requirement but a default unless user sets their own. Would be nice if we could make an unsigned request through the `request` abstraction.

Reply:
Just to confirm I understand correctly (still learning my way around both elixir and this library), you'd like to use the request function from [lib/ex_aws/request.ex](https://github.com/james-gibson-fr/ex_aws/blob/main/lib/ex_aws/request.ex#L14-L23) but because that function expects the config to have the credentials in it (which is what we're trying to get in the first place here) we can't use the function as is?

Reply:
No worries me too :D Yes, I believe so specifically when it went to add and sign headers, trying to refresh memory, it has been a while. It is almost like we need a bypass for that part or maybe there is a feature in the codebase I just missed that could be used.

Reply:
I wrote out a proposed solution for this, please check out https://github.com/joeybaer/ex_aws/pull/1/commits/37fc68442675f6822c69ead33c58d9a68a0d1315, I based it off all our recent updates.